File: src/Microsoft.CodeAnalysis.NetAnalyzers/src/Microsoft.CodeAnalysis.VisualBasic.NetAnalyzers/Microsoft.NetFramework.Analyzers/Helpers/SyntaxNodeHelper.vb
' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.

Imports Microsoft.NetFramework.Analyzers.Helpers
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Microsoft.NetFramework.VisualBasic.Analyzers.Helpers

    Public NotInheritable Class BasicSyntaxNodeHelper
        Inherits SyntaxNodeHelper

        Public Shared ReadOnly Property DefaultInstance As BasicSyntaxNodeHelper = New BasicSyntaxNodeHelper()

        Private Sub New()
        End Sub

        Public Overrides Function GetClassDeclarationTypeSymbol(node As SyntaxNode, semanticModel As SemanticModel) As ITypeSymbol
            If node Is Nothing Then
                Return Nothing
            End If

            Dim kind As SyntaxKind = node.Kind()
            If kind = SyntaxKind.ClassBlock Then
                Return semanticModel.GetDeclaredSymbol(CType(node, ClassBlockSyntax))
            End If

            Return Nothing
        End Function

        Public Overrides Function GetAssignmentLeftNode(node As SyntaxNode) As SyntaxNode
            If node Is Nothing Then
                Return Nothing
            End If

            Dim kind As SyntaxKind = node.Kind()
            If kind = SyntaxKind.SimpleAssignmentStatement Then
                Return CType(node, AssignmentStatementSyntax).Left
            End If

            If kind = SyntaxKind.VariableDeclarator Then
                Return CType(node, VariableDeclaratorSyntax).Names.First()
            End If

            If kind = SyntaxKind.NamedFieldInitializer Then
                Return CType(node, NamedFieldInitializerSyntax).Name
            End If

            Return Nothing
        End Function

        Public Overrides Function GetAssignmentRightNode(node As SyntaxNode) As SyntaxNode
            If node Is Nothing Then
                Return Nothing
            End If

            Dim kind As SyntaxKind = node.Kind()
            If kind = SyntaxKind.SimpleAssignmentStatement Then
                Return CType(node, AssignmentStatementSyntax).Right
            End If

            If kind = SyntaxKind.VariableDeclarator Then
                Dim decl As VariableDeclaratorSyntax = CType(node, VariableDeclaratorSyntax)

                If decl.Initializer IsNot Nothing Then
                    Return decl.Initializer.Value
                End If

                If decl.AsClause IsNot Nothing Then
                    Return decl.AsClause
                End If
            End If

            If kind = SyntaxKind.NamedFieldInitializer Then
                Return CType(node, NamedFieldInitializerSyntax).Expression
            End If

            Return Nothing
        End Function

        Public Overrides Function GetMemberAccessExpressionNode(node As SyntaxNode) As SyntaxNode
            If node Is Nothing Then
                Return Nothing
            End If

            Dim kind As SyntaxKind = node.Kind()
            If kind = SyntaxKind.SimpleMemberAccessExpression Then
                Return CType(node, MemberAccessExpressionSyntax).Expression
            End If

            Return Nothing
        End Function

        Public Overrides Function GetMemberAccessNameNode(node As SyntaxNode) As SyntaxNode
            If node Is Nothing Then
                Return Nothing
            End If

            Dim kind As SyntaxKind = node.Kind()
            If kind = SyntaxKind.SimpleMemberAccessExpression Then
                Return CType(node, MemberAccessExpressionSyntax).Name
            End If

            Return Nothing
        End Function

        Public Overrides Function GetInvocationExpressionNode(node As SyntaxNode) As SyntaxNode
            If node Is Nothing Then
                Return Nothing
            End If

            Dim kind As SyntaxKind = node.Kind()
            If kind = SyntaxKind.InvocationExpression Then
                Return CType(node, InvocationExpressionSyntax).Expression
            End If

            Return Nothing
        End Function

        Public Overrides Function GetCallTargetNode(node As SyntaxNode) As SyntaxNode
            If node Is Nothing Then
                Return Nothing
            End If

            Dim kind As SyntaxKind = node.Kind()
            If kind = SyntaxKind.InvocationExpression Then
                Dim callExpr As ExpressionSyntax = CType(node, InvocationExpressionSyntax).Expression
                Dim nameNode As SyntaxNode = GetMemberAccessNameNode(callExpr)
                If nameNode IsNot Nothing Then
                    Return nameNode
                Else
                    Return callExpr
                End If
            ElseIf kind = SyntaxKind.ObjectCreationExpression Then
                Return CType(node, ObjectCreationExpressionSyntax).Type
            End If

            Return Nothing
        End Function

        Public Overrides Function GetDefaultValueForAnOptionalParameter(declNode As SyntaxNode, paramIndex As Integer) As SyntaxNode
            Dim methodDecl = TryCast(declNode, MethodBlockBaseSyntax)
            If methodDecl Is Nothing Then
                Return Nothing
            End If

            Dim paramList As ParameterListSyntax = methodDecl.BlockStatement.ParameterList
            If paramIndex < paramList.Parameters.Count Then
                Dim equalsValueNode As EqualsValueSyntax = paramList.Parameters(paramIndex).Default
                If equalsValueNode IsNot Nothing Then
                    Return equalsValueNode.Value
                End If
            End If

            Return Nothing
        End Function

        Protected Overrides Function GetCallArgumentExpressionNodes(node As SyntaxNode, callKind As CallKinds) As IEnumerable(Of SyntaxNode)
            If node Is Nothing Then
                Return Nothing
            End If

            Dim argList As ArgumentListSyntax = Nothing
            Dim kind As SyntaxKind = node.Kind()

            If kind = SyntaxKind.InvocationExpression AndAlso ((callKind And CallKinds.Invocation) <> 0) Then
                argList = CType(node, InvocationExpressionSyntax).ArgumentList
            ElseIf (kind = SyntaxKind.ObjectCreationExpression) AndAlso ((callKind And CallKinds.ObjectCreation) <> 0) Then
                argList = CType(node, ObjectCreationExpressionSyntax).ArgumentList
            ElseIf (kind = SyntaxKind.AsNewClause) AndAlso ((callKind And CallKinds.ObjectCreation) <> 0) Then
                Dim asNewClause As AsNewClauseSyntax = CType(node, AsNewClauseSyntax)
                If asNewClause.NewExpression IsNot Nothing AndAlso asNewClause.NewExpression.Kind = SyntaxKind.ObjectCreationExpression Then
                    argList = CType(asNewClause.NewExpression, ObjectCreationExpressionSyntax).ArgumentList
                End If
            End If

            If argList IsNot Nothing Then
                Return From arg In argList.Arguments
                       Select arg.GetExpression()
            End If

            Return Enumerable.Empty(Of SyntaxNode)
        End Function

        Public Overrides Function GetObjectInitializerExpressionNodes(node As SyntaxNode) As IEnumerable(Of SyntaxNode)
            Dim empty As IEnumerable(Of SyntaxNode) = Enumerable.Empty(Of SyntaxNode)

            If node Is Nothing Then
                Return empty
            End If

            Dim objectCreationNode As ObjectCreationExpressionSyntax = node.DescendantNodesAndSelf().OfType(Of ObjectCreationExpressionSyntax)().FirstOrDefault()

            If objectCreationNode Is Nothing Then
                Return empty
            End If

            If objectCreationNode.Initializer Is Nothing Then
                Return empty
            End If

            Dim kind As SyntaxKind = objectCreationNode.Initializer.Kind()
            If kind <> SyntaxKind.ObjectMemberInitializer Then
                Return empty
            End If

            ' CA1804: Remove unused locals
            Dim initializer As ObjectMemberInitializerSyntax = CType(objectCreationNode.Initializer, ObjectMemberInitializerSyntax)
            Return From fieldInitializer In initializer.Initializers
                   Where fieldInitializer.Kind() = SyntaxKind.NamedFieldInitializer
                   Select CType(fieldInitializer, NamedFieldInitializerSyntax)
        End Function

        Public Overrides Function IsMethodInvocationNode(node As SyntaxNode) As Boolean
            If node Is Nothing Then
                Return False
            End If

            Dim kind As SyntaxKind = node.Kind()
            Return kind = SyntaxKind.InvocationExpression OrElse kind = SyntaxKind.ObjectCreationExpression
        End Function

        Public Overrides Function GetCalleeMethodSymbol(node As SyntaxNode, semanticModel As SemanticModel) As IMethodSymbol
            If node Is Nothing Then
                Return Nothing
            End If

            Dim kind As SyntaxKind = node.Kind()
            Dim symbol As ISymbol = GetReferencedSymbol(node, semanticModel)

            If symbol Is Nothing And kind = SyntaxKind.AsNewClause Then
                symbol = GetReferencedSymbol(node.ChildNodes().First(), semanticModel)
            End If

            If symbol IsNot Nothing Then
                If symbol.Kind = SymbolKind.Method Then
                    Return CType(symbol, IMethodSymbol)
                End If
            End If

            Return Nothing
        End Function
        Public Overrides Function GetCallerMethodSymbol(node As SyntaxNode, semanticModel As SemanticModel) As IMethodSymbol
            If node Is Nothing Then
                Return Nothing
            End If

            Dim declaration As MethodBlockSyntax = node.AncestorsAndSelf().OfType(Of MethodBlockSyntax)().FirstOrDefault()

            If declaration IsNot Nothing Then
                Return semanticModel.GetDeclaredSymbol(declaration)
            End If

            Dim constructor As SubNewStatementSyntax = node.AncestorsAndSelf().OfType(Of SubNewStatementSyntax)().FirstOrDefault()

            If constructor IsNot Nothing Then
                Return semanticModel.GetDeclaredSymbol(constructor)
            End If

            Return Nothing
        End Function

        Public Overrides Function GetEnclosingTypeSymbol(node As SyntaxNode, semanticModel As SemanticModel) As ITypeSymbol
            If node Is Nothing Then
                Return Nothing
            End If

            Dim declaration As ClassBlockSyntax = node.AncestorsAndSelf().OfType(Of ClassBlockSyntax)().FirstOrDefault()

            If declaration Is Nothing Then
                Return Nothing
            End If

            Return semanticModel.GetDeclaredSymbol(declaration)
        End Function

        Public Overrides Function GetDescendantAssignmentExpressionNodes(node As SyntaxNode) As IEnumerable(Of SyntaxNode)
            Dim empty As IEnumerable(Of SyntaxNode) = Enumerable.Empty(Of SyntaxNode)()

            If node Is Nothing Then
                Return empty
            End If

            Return node.DescendantNodesAndSelf.OfType(Of AssignmentStatementSyntax)()
        End Function

        Public Overrides Function GetDescendantMemberAccessExpressionNodes(node As SyntaxNode) As IEnumerable(Of SyntaxNode)
            Dim empty As IEnumerable(Of SyntaxNode) = Enumerable.Empty(Of SyntaxNode)()

            If node Is Nothing Then
                Return empty
            End If

            Return node.DescendantNodesAndSelf().OfType(Of MemberAccessExpressionSyntax)()
        End Function

        Public Overrides Function IsObjectCreationExpressionUnderFieldDeclaration(node As SyntaxNode) As Boolean
            Return node IsNot Nothing And node.Kind() = SyntaxKind.ObjectCreationExpression _
                And node.AncestorsAndSelf().OfType(Of FieldDeclarationSyntax)().FirstOrDefault() IsNot Nothing
        End Function

        Public Overrides Function GetVariableDeclaratorOfAFieldDeclarationNode(objectCreationExpression As SyntaxNode) As SyntaxNode
            If IsObjectCreationExpressionUnderFieldDeclaration(objectCreationExpression) Then
                Return objectCreationExpression.AncestorsAndSelf().OfType(Of VariableDeclaratorSyntax)().FirstOrDefault()
            Else
                Return Nothing
            End If
        End Function
    End Class
End Namespace

File: src/Microsoft.CodeAnalysis.NetAnalyzers/src/Microsoft.CodeAnalysis.CSharp.NetAnalyzers/Microsoft.NetFramework.Analyzers/Helpers/SyntaxNodeHelper.cs
// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.

using System.Collections.Generic;
using System.Linq;
using Microsoft.NetFramework.Analyzers.Helpers;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using System.Diagnostics.CodeAnalysis;

namespace Microsoft.NetFramework.CSharp.Analyzers.Helpers
{
    public sealed class CSharpSyntaxNodeHelper : SyntaxNodeHelper
    {
        public static CSharpSyntaxNodeHelper Default { get; } = new CSharpSyntaxNodeHelper();

        private CSharpSyntaxNodeHelper()
        { }

        public override ITypeSymbol? GetClassDeclarationTypeSymbol(SyntaxNode? node, SemanticModel semanticModel)
        {
            if (node == null)
            {
                return null;
            }

            SyntaxKind kind = node.Kind();
            if (kind == SyntaxKind.ClassDeclaration)
            {
                return semanticModel.GetDeclaredSymbol((ClassDeclarationSyntax)node);
            }

            return null;
        }

        public override SyntaxNode? GetAssignmentLeftNode(SyntaxNode? node)
        {
            if (node == null)
            {
                return null;
            }

            SyntaxKind kind = node.Kind();
            if (kind == SyntaxKind.SimpleAssignmentExpression)
            {
                return ((AssignmentExpressionSyntax)node).Left;
            }

            if (kind == SyntaxKind.VariableDeclarator)
            {
                return (VariableDeclaratorSyntax)node;
            }

            return null;
        }

        public override SyntaxNode? GetAssignmentRightNode(SyntaxNode? node)
        {
            if (node == null)
            {
                return null;
            }

            SyntaxKind kind = node.Kind();
            if (kind == SyntaxKind.SimpleAssignmentExpression)
            {
                return ((AssignmentExpressionSyntax)node).Right;
            }

            if (kind == SyntaxKind.VariableDeclarator)
            {
                EqualsValueClauseSyntax initializer = ((VariableDeclaratorSyntax)node).Initializer;
                if (initializer != null)
                {
                    return initializer.Value;
                }
            }

            return null;
        }

        public override SyntaxNode? GetMemberAccessExpressionNode(SyntaxNode? node)
        {
            if (node == null)
            {
                return null;
            }

            SyntaxKind kind = node.Kind();
            if (kind == SyntaxKind.SimpleMemberAccessExpression)
            {
                return ((MemberAccessExpressionSyntax)node).Expression;
            }

            return null;
        }

        public override SyntaxNode? GetMemberAccessNameNode(SyntaxNode? node)
        {
            if (node == null)
            {
                return null;
            }

            SyntaxKind kind = node.Kind();
            if (kind == SyntaxKind.SimpleMemberAccessExpression)
            {
                return ((MemberAccessExpressionSyntax)node).Name;
            }

            return null;
        }

        public override SyntaxNode? GetInvocationExpressionNode(SyntaxNode? node)
        {
            if (node == null)
            {
                return null;
            }

            SyntaxKind kind = node.Kind();
            if (kind != SyntaxKind.InvocationExpression)
            {
                return null;
            }

            return ((InvocationExpressionSyntax)node).Expression;
        }

        public override SyntaxNode? GetCallTargetNode(SyntaxNode? node)
        {
            if (node != null)
            {
                SyntaxKind kind = node.Kind();
                if (kind == SyntaxKind.InvocationExpression)
                {
                    ExpressionSyntax callExpr = ((InvocationExpressionSyntax)node).Expression;
                    return GetMemberAccessNameNode(callExpr) ?? callExpr;
                }
                else if (kind == SyntaxKind.ObjectCreationExpression)
                {
                    return ((ObjectCreationExpressionSyntax)node).Type;
                }
            }

            return null;
        }

        public override SyntaxNode? GetDefaultValueForAnOptionalParameter(SyntaxNode? declNode, int paramIndex)
        {
            if (declNode is BaseMethodDeclarationSyntax methodDecl)
            {
                ParameterListSyntax paramList = methodDecl.ParameterList;
                if (paramIndex < paramList.Parameters.Count)
                {
                    EqualsValueClauseSyntax equalsValueNode = paramList.Parameters[paramIndex].Default;
                    if (equalsValueNode != null)
                    {
                        return equalsValueNode.Value;
                    }
                }
            }
            return null;
        }

        protected override IEnumerable<SyntaxNode> GetCallArgumentExpressionNodes(SyntaxNode? node, CallKinds callKind)
        {
            if (node != null)
            {
                ArgumentListSyntax? argList = null;
                SyntaxKind kind = node.Kind();
                if ((kind == SyntaxKind.InvocationExpression) && ((callKind & CallKinds.Invocation) != 0))
                {
                    var invocationNode = (InvocationExpressionSyntax)node;
                    argList = invocationNode.ArgumentList;
                }
                else if ((kind == SyntaxKind.ObjectCreationExpression) && ((callKind & CallKinds.ObjectCreation) != 0))
                {
                    var invocationNode = (ObjectCreationExpressionSyntax)node;
                    argList = invocationNode.ArgumentList;
                }
                if (argList != null)
                {
                    return argList.Arguments.Select(arg => arg.Expression);
                }
            }

            return Enumerable.Empty<SyntaxNode>();
        }

        public override IEnumerable<SyntaxNode> GetObjectInitializerExpressionNodes(SyntaxNode? node)
        {
            IEnumerable<SyntaxNode> empty = Enumerable.Empty<SyntaxNode>();
            if (node == null)
            {
                return empty;
            }

            SyntaxKind kind = node.Kind();
            if (kind != SyntaxKind.ObjectCreationExpression)
            {
                return empty;
            }

            var objectCreationNode = (ObjectCreationExpressionSyntax)node;
            if (objectCreationNode.Initializer == null)
            {
                return empty;
            }

            return objectCreationNode.Initializer.Expressions;
        }

        public override bool IsMethodInvocationNode(SyntaxNode? node)
        {
            if (node == null)
            {
                return false;
            }
            SyntaxKind kind = node.Kind();
            return kind is SyntaxKind.InvocationExpression or SyntaxKind.ObjectCreationExpression;
        }

        public override IMethodSymbol? GetCalleeMethodSymbol(SyntaxNode? node, SemanticModel semanticModel)
        {
            ISymbol? symbol = GetReferencedSymbol(node, semanticModel);

            if (symbol != null && symbol.Kind == SymbolKind.Method)
            {
                return (IMethodSymbol)symbol;
            }

            return null;
        }

        public override IMethodSymbol? GetCallerMethodSymbol(SyntaxNode? node, SemanticModel semanticModel)
        {
            if (node == null)
            {
                return null;
            }

            MethodDeclarationSyntax declaration = node.AncestorsAndSelf().OfType<MethodDeclarationSyntax>().FirstOrDefault();
            if (declaration != null)
            {
                return semanticModel.GetDeclaredSymbol(declaration);
            }

            ConstructorDeclarationSyntax contructor = node.AncestorsAndSelf().OfType<ConstructorDeclarationSyntax>().FirstOrDefault();
            if (contructor != null)
            {
                return semanticModel.GetDeclaredSymbol(contructor);
            }

            return null;
        }

        public override ITypeSymbol? GetEnclosingTypeSymbol(SyntaxNode? node, SemanticModel semanticModel)
        {
            if (node == null)
            {
                return null;
            }

            ClassDeclarationSyntax declaration = node.AncestorsAndSelf().OfType<ClassDeclarationSyntax>().FirstOrDefault();

            if (declaration == null)
            {
                return null;
            }

            return semanticModel.GetDeclaredSymbol(declaration);
        }

        public override IEnumerable<SyntaxNode> GetDescendantAssignmentExpressionNodes(SyntaxNode? node)
        {
            IEnumerable<SyntaxNode> empty = Enumerable.Empty<SyntaxNode>();
            if (node == null)
            {
                return empty;
            }

            return node.DescendantNodesAndSelf().OfType<AssignmentExpressionSyntax>();
        }

        public override IEnumerable<SyntaxNode> GetDescendantMemberAccessExpressionNodes(SyntaxNode? node)
        {
            IEnumerable<SyntaxNode> empty = Enumerable.Empty<SyntaxNode>();
            if (node == null)
            {
                return empty;
            }

            return node.DescendantNodesAndSelf().OfType<MemberAccessExpressionSyntax>();
        }

        public override bool IsObjectCreationExpressionUnderFieldDeclaration([NotNullWhen(returnValue: true)] SyntaxNode? node)
        {
            return node != null &&
                   node.Kind() == SyntaxKind.ObjectCreationExpression &&
                   node.AncestorsAndSelf().OfType<FieldDeclarationSyntax>().FirstOrDefault() != null;
        }

        public override SyntaxNode? GetVariableDeclaratorOfAFieldDeclarationNode(SyntaxNode? objectCreationExpression)
        {
            if (IsObjectCreationExpressionUnderFieldDeclaration(objectCreationExpression))
            {
                return objectCreationExpression.AncestorsAndSelf().OfType<VariableDeclaratorSyntax>().FirstOrDefault();
            }
            else
            {
                return null;
            }
        }
    }
}
